utils/writeable/src/write_or_ref.rs

```diff
@@ -0,0 +1,203 @@
+/// [`Writeable::write_to_string`].
+///
+/// # Examples
+///
+/// Basic usage and behavior:
+///
+/// ```
+/// use std::fmt;
+/// use std::borrow::Cow;
+/// use writeable::Writeable;
+///
+/// struct WelcomeMessage<'s> {
+///     pub name: &'s str,
+/// }
+///
+/// impl<'s> Writeable for WelcomeMessage<'s> {
+///     // see impl in Writeable docs
+/// #    fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+/// #        sink.write_str("Hello, ")?;
+/// #        sink.write_str(self.name)?;
+/// #        sink.write_char('!')?;
+/// #        Ok(())
+/// #    }
+/// }
+///
+/// let message = WelcomeMessage { name: "Alice" };
+///
+/// assert!(matches!(
+///     writeable::write_or_ref(&message, b""),
+///     Cow::Owned(s) if s == "Hello, Alice!"
+/// ));
+/// assert!(matches!(
+///     writeable::write_or_ref(&message, b"Hello"),
+///     Cow::Owned(s) if s == "Hello, Alice!"
+/// ));
+/// assert!(matches!(
+///     writeable::write_or_ref(&message, b"Hello, Bob!"),
+///     Cow::Owned(s) if s == "Hello, Alice!"
+/// ));
+/// assert!(matches!(
+///     writeable::write_or_ref(&message, b"Hello, Alice!"),
+///     Cow::Borrowed("Hello, Alice!")
+/// ));
+///
+/// // Junk at the end is ignored:
+/// assert!(matches!(
+///     writeable::write_or_ref(&message, b"Hello, Alice!..\xFF\x00\xFF"),
+///     Cow::Borrowed("Hello, Alice!")
+/// ));
+/// ```
+///
+/// Example use case: a function that transforms a string to lowercase.
+/// We are also able to write a more efficient implementation of
+/// [`Writeable::write_to_string`] in this situation.
+///
+/// ```
+/// use std::fmt;
+/// use std::borrow::Cow;
+/// use writeable::Writeable;
+///
+/// struct MakeAsciiLower<'a>(&'a str);
+///
+/// impl<'a> Writeable for MakeAsciiLower<'a> {
+///     fn write_to<W: fmt::Write + ?Sized>(&self, sink: &mut W) -> fmt::Result {
+///         for c in self.0.chars() {
+///             sink.write_char(c.to_ascii_lowercase())?;
+///         }
+///         Ok(())
+///     }
+///     #[inline]
+///     fn write_to_string(&self) -> Cow<str> {
+///         writeable::write_or_ref(self, self.0.as_bytes())
+///     }
+/// }
+///
+/// fn make_lowercase(input: &str) -> Cow<str> {
+///     let writeable = MakeAsciiLower(input);
+///     writeable::write_or_ref(&writeable, input.as_bytes())
+/// }
+///
+/// assert!(matches!(
+///     make_lowercase("this is lowercase"),
+///     Cow::Borrowed("this is lowercase")
+/// ));
+/// assert!(matches!(
+///     make_lowercase("this is UPPERCASE"),
+///     Cow::Owned(s) if s == "this is uppercase"
+/// ));
+///
+/// assert!(matches!(
+///     MakeAsciiLower("this is lowercase").write_to_string(),
+///     Cow::Borrowed("this is lowercase")
+/// ));
+/// assert!(matches!(
+///     MakeAsciiLower("this is UPPERCASE").write_to_string(),
+///     Cow::Owned(s) if s == "this is uppercase"
+/// ));
+/// ```
+pub fn write_or_ref<'a>(writeable: &impl Writeable, reference_bytes: &'a [u8]) -> Cow<'a, str> {
```

Comment:
I'm not happy with this name. Usually methods called `write` write to one of the arguments, and don't do any allocation themselves, like this does. I think `write_to_string_or_borrow` would be a more accurate name, and would align more with `write_to_string`, which is meant to use it.